lib/validate.js

```diff
@@ -105,34 +105,35 @@ var Vp = Validator.prototype;
 // (AST, string[, Function]) -> any
 // Delegates to pattern-match library, converting MatchErrors to validation errors.
 Vp.match = function(node, desc, body) {
+    var self = this;
     if (body) {
         try {
             return match(node, body, this);
         } catch (e) {
-            if (e instanceof match.MatchError)
-                this.fail("invalid " + desc);
-            else
-                throw e;
+            failWithLocation(e);
         }
     }
 
     var delayed = match(node);
-    var self = this;
 
     return {
         when: function(pattern, template) {
             try {
                 return delayed.when(pattern, template, self);
             } catch (e) {
-                if (e instanceof match.MatchError) {
-                    var loc = e && e.actual ? e.actual.loc : null;
-                    self.fail("invalid " + desc + fail.locToString(loc), loc);
-                } else {
-                    throw e;
-                }
+                failWithLocation(e);
             }
         }
     };
+
+    function failWithLocation(e) {
+        if (e instanceof match.MatchError) {
+            var loc = e && e.actual ? e.actual.loc : null;
+            self.fail("invalid " + desc + fail.locToString(loc), loc);
```

Comment:
It seems you're right. And actually, I think it's worth fixing before merging it, because it causes a different regression -- locations for errors in the body are printed twice with this patch.